File: android/app/src/main/kotlin/io/golioth/pouchprov/app/SettingsScreen.kt
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

package io.golioth.pouchprov.app

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.unit.dp
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Composable
fun SettingsScreen(vm: ProvViewModel) {
    val apiKey by vm.goliothApiKey.collectAsState()
    val validityDays by vm.certValidityDays.collectAsState()

    Column(Modifier.fillMaxWidth()) {
        // -- Golioth --------------------------------------------------------
        Text("Golioth", style = MaterialTheme.typography.titleMedium)
        var showKey by remember { mutableStateOf(false) }
        OutlinedTextField(
            value = apiKey,
            onValueChange = { vm.setGoliothApiKey(it) },
            label = { Text("Project API key") },
            singleLine = true,
            visualTransformation = if (showKey) VisualTransformation.None else PasswordVisualTransformation(),
            trailingIcon = {
                TextButton(onClick = { showKey = !showKey }) { Text(if (showKey) "Hide" else "Show") }
            },
            modifier = Modifier.fillMaxWidth().padding(top = 8.dp),
        )
        Text(
            "Used to register a temporary CA when minting device certificates.",
            style = MaterialTheme.typography.bodySmall,
            modifier = Modifier.padding(top = 4.dp),
        )

        // -- Certificate ----------------------------------------------------
        Text(
            "Certificate",
            style = MaterialTheme.typography.titleMedium,
            modifier = Modifier.padding(top = 24.dp),
        )
        var daysText by remember(validityDays) { mutableStateOf(validityDays.toString()) }
        OutlinedTextField(
            value = daysText,
            onValueChange = { input ->
                daysText = input.filter { it.isDigit() }.take(4)
                daysText.toIntOrNull()?.let { vm.setCertValidityDays(it) }
            },
            label = { Text("Expiration (days from creation)") },
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            modifier = Modifier.fillMaxWidth().padding(top = 8.dp),
        )
        val expiry = remember(validityDays) {
            LocalDate.now().plusDays(validityDays.toLong()).format(DateTimeFormatter.ISO_LOCAL_DATE)
        }
        Text(
            "A certificate generated today would expire $expiry.",
            style = MaterialTheme.typography.bodySmall,
            modifier = Modifier.padding(top = 4.dp),
        )
    }
}

File: ios/PouchProvApp/Sources/SettingsScreen.swift
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

import SwiftUI

struct SettingsScreen: View {
    @ObservedObject var viewModel: AppViewModel

    @State private var showKey = false
    @State private var daysText = ""

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            // MARK: Golioth
            Text("Golioth")
                .font(.headline)

            HStack {
                Group {
                    if showKey {
                        TextField("Project API key", text: apiKeyBinding)
                    } else {
                        SecureField("Project API key", text: apiKeyBinding)
                    }
                }
                .textFieldStyle(.roundedBorder)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)

                Button(showKey ? "Hide" : "Show") { showKey.toggle() }
                    .font(.footnote)
            }
            Text("Used to register a temporary CA when minting device certificates.")
                .font(.footnote)
                .foregroundStyle(.secondary)

            // MARK: Certificate
            Text("Certificate")
                .font(.headline)
                .padding(.top, 24)

            TextField("Expiration (days from creation)", text: $daysText)
                .textFieldStyle(.roundedBorder)
                .keyboardType(.numberPad)
                .onChange(of: daysText) { input in
                    let filtered = String(input.filter(\.isNumber).prefix(4))
                    if filtered != input { daysText = filtered }
                    if let days = Int(filtered) { viewModel.setCertValidityDays(days) }
                }

            Text("A certificate generated today would expire \(expiryText).")
                .font(.footnote)
                .foregroundStyle(.secondary)
        }
        .onAppear {
            daysText = String(viewModel.certValidityDays)
        }
    }

    private var apiKeyBinding: Binding<String> {
        Binding(
            get: { viewModel.goliothApiKey },
            set: { viewModel.setGoliothApiKey($0) }
        )
    }

    private var expiryText: String {
        let expiry = Calendar.current.date(
            byAdding: .day,
            value: viewModel.certValidityDays,
            to: Date()
        ) ?? Date()
        return expiry.formatted(date: .numeric, time: .omitted)
    }
}
